base/ca/src/org/dogtagpki/ct/CTEngine.java

```diff
@@ -0,0 +1,704 @@
+
+/**
+ * Certificate Transparency (v1) engine for CA
+ *   to issue CT certs with Embedded Signed Certificate Timestamp
+ *   - https://tools.ietf.org/html/rfc6962
+ *
+ * @author Christina Fu
+ *
+ */
+public class CTEngine {
+    public static org.slf4j.Logger logger =
+            org.slf4j.LoggerFactory.getLogger(CTEngine.class);
+
+    SCTProcessor ctConfig = null;
+    public void init()  throws Exception {
+        // for getting CT config from CS.cfg
+        ctConfig = new SCTProcessor();
+
+        // Initialize CT configuration
+        ctConfig.init();
+    }
+
+    /**
+     * Process cert info for Certificate Transparency
+     * 
+     * Check to see if certInfo contains Certificate Transparency poison
+     * extension (from profile containig certTransparencyExtDefaultImpl);
+     *
+     * If not, check if global setting is enabled in CS.cfg:
+     *    ca.certTransparency.enable
+     * and add the poision extension;
+     *
+     * if it does then reach out to the CT log servers to obtain
+     * signed certificate timestamp (SCT) for inclusion of the SCT extension
+     * in the cert to be issued.
+     *
+     */
+    public void process(X509CertInfo certi, ICertificateAuthority ctCA, AuthorityID aid, String algname)
+            throws EBaseException {
+
+        String method = "CTEngine.process: ";
+        ICertificateAuthority ca = ctCA.getCA(aid);
+        String errMsg = "";
+
+        try {
+            if (ctConfig == null) {
+                init();
+            }
+
+            CertificateExtensions exts = (CertificateExtensions) certi.get(X509CertInfo.EXTENSIONS);
+            logger.debug(method + " about to check CT poison");
+            Extension ctPoison = null;
+            try {
+                ctPoison = (Extension) exts.get("1.3.6.1.4.1.11129.2.4.3");
+                logger.debug(method + " CT poison extension found");
+            } catch (Exception e) {
+                logger.debug(method + e.getMessage() + "-- continue");
+                logger.debug(method + " CT poison extension not found");
+            }
+
+            /*
+             * see CTmode for CT mode config info
+             */
+            boolean processCT = false;
+            SCTProcessor.CTmode ct_mode = ctConfig.getCTmode();
+            if (ct_mode == SCTProcessor.CTmode.enabled) {
+                logger.debug(method + "ct_mode is enabled");
+                if (ctPoison == null) {
+                    logger.debug(method + " adding poison ext");
+                    CertUtils.addCTv1PoisonExt(certi);
+                    logger.debug(method + " returned from addCTpoisonExt");
+                }
+                processCT = true;
+            } else if (ct_mode == SCTProcessor.CTmode.perProfile) {
+                logger.debug(method + "ct_mode is perProfile");
+                if (ctPoison != null)
+                    processCT = true;
+            } else { // ct_mode == SCTProcessor.CTmode.disabled
```

Comment:
ok better done in getCTmode() instead as ct_mode simply gets reuturn from that.

Reply:
so, I decided to be less forgiving on this.  So instead of silently setting the mode to enabled or disabled, I'll juts throw an exception and let it fail so the admin could make a conscious decision on what is intended.